cni/network/invoker_cns.go

```diff
@@ -0,0 +1,78 @@
+package network
+
+import (
+	"encoding/json"
+	"net"
+
+	"github.com/Azure/azure-container-networking/cni"
+	"github.com/Azure/azure-container-networking/cns"
+	"github.com/Azure/azure-container-networking/cns/cnsclient"
+	"github.com/Azure/azure-container-networking/network"
+	cniSkel "github.com/containernetworking/cni/pkg/skel"
+	cniTypesCurr "github.com/containernetworking/cni/pkg/types/current"
+)
+
+type CNSInvoker struct {
+	podName      string
+	podNamespace string
+	cnsClient    *cnsclient.CNSClient
+}
+
+func NewCNSInvoker(podName, namespace string) (*CNSInvoker, error) {
+	cnsClient, err := cnsclient.GetCnsClient()
+	return &CNSInvoker{
+		podName:      podName,
+		podNamespace: namespace,
+		cnsClient:    cnsClient,
+	}, err
+}
+
+func (invoker *CNSInvoker) Add(args *cniSkel.CmdArgs, nwCfg *cni.NetworkConfig, nwInfo network.NetworkInfo, isDeletePoolOnError bool) (*cniTypesCurr.Result, *cniTypesCurr.Result, error) {
+	var (
+		result   *cniTypesCurr.Result
+		resultV6 *cniTypesCurr.Result
+	)
+
+	cnsclient.InitCnsClient(nwCfg.CNSUrl)
+	client, err := cnsclient.GetCnsClient()
+
+	// Parse Pod arguments.
+	podInfo := cns.KubernetesPodInfo{PodName: invoker.podName, PodNamespace: invoker.podNamespace}
+	orchestratorContext, err := json.Marshal(podInfo)
+
+	response, err := client.RequestIPAddress(orchestratorContext)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// set result ipconfig from CNS Response Body
+	ip, ipnet, err := response.IPConfiguration.IPSubnet.GetIPNet()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// construct ipnet for result
+	resultIPnet := net.IPNet{
+		IP:   ip,
+		Mask: ipnet.Mask,
+	}
+
+	result.IPs = make([]*cniTypesCurr.IPConfig, 1)
+	result.IPs[0] = &cniTypesCurr.IPConfig{
+		Address: resultIPnet,
+	}

```

Comment:
instead of doing this way, can we initialize cniTypesCurr.IPConfig and then append to array..so that it can be extensible

```
ipconfig := &cniTypesCurr.IPConfig{}
var ipconfigs []*cniTypesCurr.IPConfig
ipconfigs = append(ipconfigs, ipconfig)
```

Reply:
updating to add gwip and dns